### PR TITLE
Add missing definitions used with debug build

### PR DESCRIPTION
**Pull request by dlaugt**
There are some compilation errors when py2exe is built in debug mode:
```
python_d setup.py build --debug
...
   Creating library build\temp.win-amd64-cpython-310-pydebug\Debug\source\run_d-py3.10-win-amd64.lib and object build\temp.win-amd64-cpython-310-pydebug\Debug\source\run_d-py3.10-win-amd64.exp
start.obj : error LNK2001: unresolved external symbol _Py_NegativeRefcount
MyLoadLibrary.obj : error LNK2001: unresolved external symbol _Py_NegativeRefcount
_memimporter.obj : error LNK2001: unresolved external symbol _Py_NegativeRefcount
start.obj : error LNK2001: unresolved external symbol _Py_RefTotal
MyLoadLibrary.obj : error LNK2001: unresolved external symbol _Py_RefTotal
_memimporter.obj : error LNK2001: unresolved external symbol _Py_RefTotal
build\lib.win-amd64-cpython-310-pydebug\py2exe\run_d-py3.10-win-amd64.exe : fatal error LNK1120: 2 unresolved externals
```
Added the missing definitions taken from object.h:
```cpp
#ifdef Py_REF_DEBUG
PyAPI_DATA(Py_ssize_t) _Py_RefTotal;
PyAPI_FUNC(void) _Py_NegativeRefcount(const char *filename, int lineno,
                                      PyObject *op);
#endif /* Py_REF_DEBUG */
```

### CODE REVIEW

**Comment by albertosottile** [comment]
Hi, thank you for this contribution. Debug builds are something that I inherited from the original code and never questioned. Usually, I create them via changing this if in `setup.py`

https://github.com/py2exe/py2exe/blob/43ee68b0fdf474dcedb1ee68fb1445f1d5a9b023/setup.py#L53-L58

then building as usual with `python setup.py bdist_wheel`

If I do so, I obtain a `cpXX` tagged wheel (no `d`) with debug symbols for `py2exe` inside. 

I tried to run `python_d setup.py bdist_wheel` on my build environment and it builds just fine even without this PR (when the `1` flag is toggled in `setup.py`). With this command, I get a `cp39d` wheel, but the content is identical to the one obtained before (debug symbols for `py2exe` inside}. However, I guess the `d` wheel will not be installed by `pip` in regular Python installations, hence the need for both (though usually I debug by freezing on release Python).

I then tried your command and reproduced your error. It looks like `build` and `bdist_wheel` use a different set of symbols? 

In short, I would have a few questions:
- Do you have any idea why the symbols needs are different for `build` and `bdist_wheel`?
- Do you see a use case for `build` vs `bdist_wheel` for your debugging needs?
- Is it possible to detect from `setup.py` if we are running from `python` or `python_d`, so that the `if` condition can be extended to something like `if TOGGLE_DEBUG or PYTHON_D_IS_RUNNING`?

**Comment by dlaugt** [comment]
Hi,

I guess the trick done in setup.py to compile in debug mode doesn't work. When I run `python_d setup.py bdist_wheel` with the `1` flag is toggled in `setup.py`, I see the following compilation line:
`cl.exe /c /nologo /O2 /W3 /GL /DNDEBUG ...`

The define `NDEBUG` is passed to the compiler to say that we are in release mode. One way to fix the trick can be to give also the define `_DEBUG`:
```cpp
if 1:
    # enable this to debug a release build
    extra_compile_args.append("/Od")
    extra_compile_args.append("/Z7")
    extra_compile_args.append("/D_DEBUG")
    extra_link_args.append("/DEBUG")
    macros.append(("VERBOSE", "1"))
```
With this, you can reproduce the compilation errors. However, I would say that this trick is not very good because:
  - in the command line, there are two defines: `NDEBUG` and `_DEBUG`,
  - the tool still believes you want to compile in release mode. For example, the object files are in a Release directory (build\temp.win-amd64-cpython-310-pydebug\Release).

I have never used yet the command `bdist_wheel`. However, as far as I see, the command `bdist_wheel` is calling internally the command `build`. So, in the command line, you could give the two commands to specify the build option of the command `bdist_wheel`:
`python_d setup.py bdist_wheel build --debug`

From this, the command `bdist_wheel` will call the command `build` with option `--debug`.

Note:
```
python_d setup.py bdist_wheel --help
Options for 'bdist_wheel' command:
  --bdist-dir (-b)  temporary directory for creating the distribution
  --plat-name (-p)  platform name to embed in generated filenames (default:
                    win_amd64)
  --keep-temp (-k)  keep the pseudo-installation tree around after creating
                    the distribution archive
  --dist-dir (-d)   directory to put final built distributions in
  --skip-build      skip rebuilding everything (for testing/debugging)
  --relative        build the archive using relative paths (default: false)
  --owner (-u)      Owner name used when creating a tar file [default: current
                    user]
  --group (-g)      Group name used when creating a tar file [default: current
                    group]
  --universal       make a universal wheel (default: false)
  --compression     zipfile compression (one of: stored, deflated) (default:
                    'deflated')
  --python-tag      Python implementation compatibility tag (default: 'py3')
  --build-number    Build number for this particular version. As specified in
                    PEP-0427, this must start with a digit. [default: None]
  --py-limited-api  Python tag (cp32|cp33|cpNN) for abi3 wheel tag (default:
                    false)
```
```
python_d setup.py build --help
Options for 'build' command:
  --build-base (-b)  base directory for build library
  --build-purelib    build directory for platform-neutral distributions
  --build-platlib    build directory for platform-specific distributions
  --build-lib        build directory for all distribution (defaults to either
                     build-purelib or build-platlib
  --build-scripts    build directory for scripts
  --build-temp (-t)  temporary build directory
  --plat-name (-p)   platform name to build for, if supported (default: win-
                     amd64)
  --compiler (-c)    specify the compiler type
  --parallel (-j)    number of parallel build jobs
  --debug (-g)       compile extensions and libraries with debugging
                     information
  --force (-f)       forcibly build everything (ignore file timestamps)
  --executable (-e)  specify final destination interpreter path (build.py)
  --help-compiler    list available compilers
```

**Comment by albertosottile** [comment]
Understood. So we have two use cases for debug builds: 
- create wheels linked to the Python debug build (e.g `cp39d`) with debug symbols for `py2exe`, which will be covered by `python(_d) setup.py bdist_wheel build --debug` after this MR
- create wheels linked to the Python release build (e.g. `cp39`) with debug symbols for `py2exe`.

I have the second use case because I have to run tests on release Python (to be able to test some C extensions from PyPI, e.g. `numpy`, `scipy`, ...) but sometimes I still need `py2exe` debug symbols. 

My question would be: can I address the second use case without "the trick done in setup.py"? If so, I would be glad to integrate the changes needed to do this properly in this PR, and change my workflow accordingly. If not, then I guess I would just merge this PR as it is but keep the "hack if" in there due to the lack of better alternatives.

**Comment by dlaugt** [comment]
I understand your point of debugging a release build. In this use case, the "hack if" is good and you can keep it. For Windows builds, I don't see any other way to pass these custom compiler and linker flags.

**Comment by albertosottile** [comment]
Thank you again for this contribution!
